Sources/Vapor/HTTP/Server/HTTPServerResponseEncoder.swift

```diff
@@ -103,41 +102,42 @@ private final class ChannelResponseBodyStream: BodyStreamWriter {
         promise: EventLoopPromise<Void>?,
         count: Int?
     ) {
-        self.context = context
-        self.handler = handler
+        self.contextBox = .init(context, eventLoop: context.eventLoop)
+        self.handlerBox = .init(handler, eventLoop: context.eventLoop)
         self.promise = promise
         self.count = count
-        self.currentCount = 0
-        self.isComplete = false
+        self.currentCount = .init(0)
+        self.isComplete = .init(false)
+        self.eventLoop = context.eventLoop
     }
     
     func write(_ result: BodyStreamResult, promise: EventLoopPromise<Void>?) {
         switch result {
         case .buffer(let buffer):
-            self.context.writeAndFlush(self.handler.wrapOutboundOut(.body(.byteBuffer(buffer))), promise: promise)
-            self.currentCount += buffer.readableBytes
-            if let count = self.count, self.currentCount > count {
+            self.contextBox.value.writeAndFlush(self.handlerBox.value.wrapOutboundOut(.body(.byteBuffer(buffer))), promise: promise)
+            self.currentCount.wrappingIncrement(by: buffer.readableBytes, ordering: .sequentiallyConsistent)
+            if let count = self.count, self.currentCount.load(ordering: .sequentiallyConsistent) > count {
                 self.promise?.fail(Error.tooManyBytes)
                 promise?.fail(Error.notEnoughBytes)
             }
         case .end:
-            self.isComplete = true
-            if let count = self.count, self.currentCount != count {
+            self.isComplete.store(true, ordering: .relaxed)
+            if let count = self.count, self.currentCount.load(ordering: .sequentiallyConsistent) != count {
```

Comment:
```suggestion
            if let count = self.count, self.currentCount.load(ordering: .sequentiallyConsistent) < count {
```
Preexisting semantic bug - "too many bytes" is not the same as "not enough bytes".

Reply:
Ok I need to dig into this, because we should probably have a test for it